File: deep_learning/game/slidepuzzle.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod
from .move import Move
from .piece import Piece

class SlidePuzzle(ABC):

    def __init__(self):
        self.__last_move = Move.STAY
        self.__move_count = 0

    @property
    def last_move(self):
        return self.__last_move

    @last_move.setter
    def last_move(self, move):
        self.__last_move = move

    @property
    def move_count(self):
        return self.__move_count

    def __inc_move_count(self):
        ++self.__move_count

    def __reset_move_count(self):
        self.__move_count = 0

    @abstractmethod
    def get_pos_blank(self):
        pass

    @abstractmethod
    def __update_pos_blank(self, delta_y, delta_x):
        pass

    @property
    @abstractmethod
    def row(self):
        pass

    @row.setter
    @abstractmethod
    def row(self, row):
        pass

    @property
    @abstractmethod
    def column(self):
        pass

    @column.setter
    @abstractmethod
    def column(self, column):
        pass

    @abstractmethod
    def get_cell(self, pos):
        pass

    @abstractmethod
    def __set_cell(self, pos, other):
        pass

    def move(self, move):
        if(self.check_move(move)):
            y, x = self.get_pos_blank()
            delta_y, delta_x = 0, 0
            if move == Move.LEFT:
                x -= 1
                delta_x = -1
            elif move == Move.RIGHT:
                x += 1
                delta_x = 1
            elif move == Move.UP:
                y -= 1
                delta_y = -1
            elif move == Move.DOWN:
                y += 1
                delta_y = 1
            else:
                return False

            self.__inc_move_count()
            self.last_move = move
            self.__swap_cell(self.get_pos_blank(), (y, x))
            self.__update_pos_blank(delta_y, delta_x)
            return True
        else:
            return False

    
    def check_move(self, move):
        switcher = {
            Move.LEFT: (0, -1),
            Move.RIGHT: (0, 1),
            Move.UP: (-1, 0),
            Move.DOWN: (1, 0)
        }

        case = switcher.get(move, Move.STAY)
        if(case != Move.STAY):
            return self.check_delta_move(case[0], case[1])
        else:
            return False

    
    def check_delta_move(self, y_move, x_move):
        return (self.get_pos_blank()[0] + y_move < self.row and
                self.get_pos_blank()[0] + y_move >= 0 and
                self.get_pos_blank()[1] + x_move < self.column and
                self.get_pos_blank()[1] + x_move >= 0)

    def __swap_cell(self, left, right):
        temp = Piece(self.get_cell(left).id,
                                 self.get_cell(left).value)
        self.__set_cell(left, self.get_cell(right))
        self.__set_cell(right, temp)

File: android/NPuzzle/app/src/main/java/com/repelliuss/npuzzle/game/SlidePuzzle.java
package com.repelliuss.npuzzle.game;

import androidx.annotation.NonNull;

import com.repelliuss.npuzzle.utils.Index2D;
import com.repelliuss.npuzzle.utils.Move;

import java.util.Random;

public abstract class SlidePuzzle<T> implements Puzzle<SlidePuzzle<T>.Piece> {

    private Move lastMove;
    private int moveCount;
    private GameStatus status;
    private final int MIN_RANDOM_MOVE = 200;
    private final int EXTRA_RANDOM_MOVE = 200;

    public enum Cell {
        VALUE,
        BLANK
    }

    public class Piece implements BoardPiece<Cell, T> {

        Cell id;
        T value;

        public Piece(Cell argID) {
            setId(argID);
        }

        public Piece(Cell argId, final T argValue) {
             setPiece(argId, argValue);
        }

        public Piece(final Piece other) {
            id = other.getId();
            value = other.getValue();
        }

        @Override
        public Cell getId() { return id; }

        @Override
        public T getValue() { return value; }

        public void setId(final Cell argId) { id = argId; }

        public void setValue(final T argValue) { value = argValue; }

        public void setPiece(final Piece other) {
            setId(other.getId());
            setValue(other.getValue());
        }

        public void setPiece(Cell argId, final T argValue) {
            setId(argId);
            setValue(argValue);
        }
    }

    protected SlidePuzzle() {

        lastMove = Move.STAY;
        moveCount = 0;
        status = GameStatus.ONGOING;
    }

    @Override
    public Move getLastMove() {
        return lastMove;
    }

    @Override
    public int getMoveCount() {
        return moveCount;
    }

    private void setLastMove(Move argLastMove) {
        lastMove = argLastMove;
    }

    private void incMoveCount() {
        ++moveCount;
    }

    private void resetMoveCount() {
        moveCount = 0;
    }

    public abstract Index2D getPosBlank();
    protected abstract void setRow(int argRow);
    protected abstract void setColumn(int argColumn);
    protected abstract void setPiece(Index2D pos, final Piece piece);

    @Override
    public GameStatus getStatus() {
        return status;
    }

    private void setStatus(GameStatus updatedStatus) {
        status = updatedStatus;
    }

    @Override
    public void initialize() {
        randomize();
    }

    @Override
    public boolean move(@NonNull final Move move) {
        if(checkMove(move)) {
            switch(move) {
                case LEFT:
                    swapCell(getPosBlank(), Index2D.getRelativeToX(getPosBlank(), -1));
                    getPosBlank().decX();
                    break;
                case RIGHT:
                    swapCell(getPosBlank(), Index2D.getRelativeToX(getPosBlank(), 1));
                    getPosBlank().incX();
                    break;
                case UP:
                    swapCell(getPosBlank(), Index2D.getRelativeToY(getPosBlank(), -1));
                    getPosBlank().decY();
                    break;
                case DOWN:
                    swapCell(getPosBlank(), Index2D.getRelativeToY(getPosBlank(), 1));
                    getPosBlank().incY();
                    break;
            }

            incMoveCount();
            setLastMove(move);
            if(isSolved())
                setStatus(GameStatus.FINISHED);
            else setStatus(GameStatus.ONGOING);

            return true;
        }

        return false;
    }

    @Override
    public boolean checkMove(@NonNull final Move move) {
        switch(move) {
            case LEFT:
                return checkMove(0, -1);
            case RIGHT:
                return checkMove(0, 1);
            case UP:
                return checkMove(-1, 0);
            case DOWN:
                return checkMove(1, 0);
            default:
                return false;
        }
    }

    @Override
    public boolean checkMove(int yMove, int xMove) {

        return  getPosBlank().getY() + yMove < getRow() &&
                getPosBlank().getY() + yMove >= 0 &&
                getPosBlank().getX() + xMove < getColumn() &&
                getPosBlank().getX() + xMove >= 0;
    }

    public void randomize() {

        Move[] moveList = new Move[]{ Move.UP, Move.DOWN, Move.LEFT, Move.RIGHT};
        Random random = new Random();
        int moveCount = MIN_RANDOM_MOVE + random.nextInt(EXTRA_RANDOM_MOVE);
        int range = moveList.length;
        int moveIndex = random.nextInt(range--);

        for(int i = 0; i < moveCount; ++i) {

            while(!move(moveList[moveIndex])) {
                Move.swap(moveList, moveIndex, range--);
                moveIndex = random.nextInt(range);
            }

            Move opposite = Move.toOpposite(moveList[moveIndex]);
            int oppositeIndex = 0;
            for(int j = 0; j < moveList.length; ++j)
                if(moveList[j] == opposite)
                    oppositeIndex = j;

            Move.swap(moveList, oppositeIndex, moveList.length - 1);
            range = moveList.length - 1;
            moveIndex = random.nextInt(range);
        }

        if(isSolved()) move(Move.LEFT);

        setLastMove(Move.STAY);
        resetMoveCount();
        setStatus(GameStatus.ONGOING);
    }

    private void swapCell(Index2D left, Index2D right) {
        Piece temp = new Piece(getPiece(left));
        setPiece(left, getPiece(right));
        setPiece(right, temp);
    }
}
